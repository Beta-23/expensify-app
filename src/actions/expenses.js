import {v4 as uuid} from 'uuid';
// ADD_EXPENSE action type generator
export const addExpense = ({description = '', note = '', amount = 0, createdAt = 0} = {}) => ({
	type: 'ADD_EXPENSE',
	expense: {
		id: uuid(),
		description,
		note,
		amount,
		createdAt
	}
});

// REMOVE_EXPENSE action type generator
export const removeExpense = ({id} = {}) => ({
	type: 'REMOVE_EXPENSE',
	id
});

// EDIT_EXPENSE action type generator
export const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
});
