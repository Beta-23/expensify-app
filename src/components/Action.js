import React from 'react';

// Action stateless functional component for User
const Action = (props) => (
        <div>
            <button
                className="big-button" 
                onClick={props.handlePickOptions}
                disabled={!props.hasOptions}
                >
                What langueage should I learn?
            </button>
        </div>
);

export default Action;