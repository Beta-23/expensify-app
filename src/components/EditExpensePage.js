import React from 'react';

const EditExpensePage = () => (
    <div>
        <h5>EDIT | App Expense Component</h5>
    </div>
);

export default EditExpensePage;