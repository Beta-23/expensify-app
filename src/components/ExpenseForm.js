import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/initialize';

export default class ExpenseForm extends React.Component {
	constructor (props) {
		super(props);
		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarfocused: false
		};
	}
	// Event listeners
	onDescriptionChange = (e) => {
		const description = e.target.value;
		this.setState(() => ({description}));
	};

	// used e.persist event listeners
	onNoteChange = (e) => {
		e.persist();
		this.setState(() => ({note: e.target.value}));
	};
	// Amount input Logic with regular expressions starting with 1 to âˆž
	onAmountChange = (e) => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(() => ({amount}));
		}
	};

	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({createdAt}));
		}
	};

	onFocusChange = ({focused}) => {
		this.setState(() => ({ calendarFocused: focused }));
		
	};

	onSubmit = (e) => {
		e.preventDefault();
		if (!this.state.description || !this.state.amount) {
			this.setState(() => ({
				error: 'Please Provide Valid Description and/or Amount'
			}));
		}
		else {
			this.setState(() => {
				error:"";
			});
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100, // convert from string to float
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			});
		}
	};

	render () {
		return (
			<div>
				{this.state.error && <p>{this.state.error}</p>}
				<form onSubmit={this.onSubmit}>
					<input
						type='text'
						placeholder='Description'
						autoFocus
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>
					<input type='text' placeholder='Amount' value={this.state.amount} onChange={this.onAmountChange} />

					<SingleDatePicker
						date={this.state.createdAt} // momentPropTypes.momentObj or null
						onDateChange={this.onDateChange} // PropTypes.func.isRequired, track changes
						focused={this.state.calendarfocused} // PropTypes.bool
						onFocusChange={this.onFocusChange} // PropTypes.func.isRequired, track changes
						numberOfMonths={1} // number of month calendars on page
						isOutsideRange={() => false} // every single day available
					/>
					<textarea placeholder='Add Expense Notes (optional)' value={this.state.note} onChange={this.onNoteChange} />
					<button>Add Expense</button>
				</form>
			</div>
		);
	}
}
