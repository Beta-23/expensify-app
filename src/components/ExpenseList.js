import React from 'react';
import {connect} from 'react-redux';

const ExpenseList = (props) => (
	<div>
		<h5>DASH | App dashboard Component</h5>
		{props.filters.text}
		{props.expenses.length}
	</div>
);

const mapStateToProps = (state) => {
	return {
		expenses: state.expenses,
		filters: state.filters
	};
};

// Using the connect library for props to use and to what component
export default connect(mapStateToProps)(ExpenseList);
