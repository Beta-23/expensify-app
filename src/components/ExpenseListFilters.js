import React from 'react';
import {connect} from 'react-redux';

const ExpenseListFilters = (props) => (
	<div>
		<input type='text' value={props.filters.text} />
	</div>
);

// Grabbing props from state
const mapStateToProps = (state) => {
	return {
		filters: state.filters
	};
};

export default connect(mapStateToProps)(ExpenseListFilters);
