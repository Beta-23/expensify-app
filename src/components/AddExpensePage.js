import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';

// Functional Components

const AddExpensePage = (props) => (
	<div>
		<h1>ADD EXPENSES</h1>
		<ExpenseForm
			onSubmit={(expense) => {
				props.dispatch(addExpense(expense));
				props.history.push('/'); // Use the history method to push data to dashboard
			}}
		/>
	</div>
);

export default connect()(AddExpensePage);
