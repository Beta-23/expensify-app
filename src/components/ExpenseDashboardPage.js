import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
// Functional Components
const ExpenseDashboardPage = () => (
	<div>
		<ExpenseListFilters />
		<ExpenseList />
	</div>
);

export default ExpenseDashboardPage;
