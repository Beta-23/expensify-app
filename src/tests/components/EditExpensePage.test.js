import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { EditExpensePage } from '../../components/EditExpensePage';

let editExpense, wrapper;

beforeEach(() => {
    editExpense = jest.fn();
    
    wrapper = shallow(
        <EditExpensePage
            editExpense={editExpense}
        />
    );
});

test('should render EditExpensePage', () => {
    expect(wrapper).toMatchSnapshot();
});


