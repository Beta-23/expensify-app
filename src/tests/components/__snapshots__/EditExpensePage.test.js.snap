// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditExpensePage
    editExpense={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ExpenseForm
          expense={undefined}
          onSubmit={[Function]}
/>,
        <button
          onClick={[Function]}
>
          Remove
</button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "expense": undefined,
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Remove",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Remove",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ExpenseForm
            expense={undefined}
            onSubmit={[Function]}
/>,
          <button
            onClick={[Function]}
>
            Remove
</button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "expense": undefined,
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Remove",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Remove",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
